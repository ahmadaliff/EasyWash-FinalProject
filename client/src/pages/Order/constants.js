export const SET_ORDER = 'Order/SET_ORDER';
export const RESET_ORDER = 'Order/RESET_ORDER';
export const CREATE_ORDER = 'order/CREATE_ORDER';
