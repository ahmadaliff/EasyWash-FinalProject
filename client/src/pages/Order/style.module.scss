@import '@styles/base';

.mainWrap {
  @include directionY;
  padding: 1rem;
  @include apply-at-md-size {
    padding: 2rem 5rem 2rem 5rem;
  }
  min-height: calc(100vh - 3.55rem);
  align-items: center;
  box-sizing: border-box;
  position: relative;
  * {
    box-sizing: border-box;
  }
  .header {
    @include directionX;
    justify-content: space-between;
    width: 100%;
    align-items: center;
    .backButton {
      margin-left: auto;
      background-color: var(--color-primary);
      border-radius: 10rem;
      padding: 0.25rem 1rem 0.25rem 1rem;
      text-transform: capitalize;
      font-size: medium;
      &:hover {
        background-color: var(--color-primary);
      }
    }
  }
  .service {
    @include directionX;
    align-items: center;
    gap: 0.125rem;
    padding: 0.25rem;
    width: 100%;
    .serviceInfoWrap {
      width: 100%;
      p {
        margin: 0;
        font-size: small;
      }
      .quantity {
        @include directionX;
        align-items: center;
        justify-content: space-between;
        font-size: small;
      }
      .serviceInfo {
        @include directionX;
        justify-content: space-between;
      }
    }
  }
  .totalPrice {
    @include directionX;
    width: 100%;
    justify-content: space-between;
    font-size: small;
    font-weight: 700;
  }
  .inputLabel {
    @include directionY;
    gap: 0.25rem;
    width: 100%;
    padding: 0.5rem 0 0.5rem 0;
    font-weight: 400;
    font-size: small;
  }
  .createButton {
    width: 100%;
    background-color: var(--color-primary);
    padding: 0.25rem 1rem 0.25rem 1rem;
    text-transform: capitalize;
    font-size: small;
    font-weight: 700;
    &:hover {
      background-color: var(--color-primary);
    }
  }
}
