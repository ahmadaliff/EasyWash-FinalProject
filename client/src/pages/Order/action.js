import { CREATE_ORDER, RESET_ORDER, SET_ORDER } from './constants';

export const actionSetOrder = (data) => ({
  type: SET_ORDER,
  data,
});

export const actionCreateOrder = (data, callback) => ({
  type: CREATE_ORDER,
  data,
  callback,
});

export const actionResetOrder = () => ({
  type: RESET_ORDER,
});
