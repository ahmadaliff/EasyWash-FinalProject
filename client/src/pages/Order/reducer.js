import { produce } from 'immer';
import { RESET_ORDER, SET_ORDER } from './constants';

export const initialState = {
  order: null,
};

export const storedKey = ['order'];

export const orderReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case SET_ORDER:
        draft.order = action.data;
        break;
      case RESET_ORDER:
        return initialState;
    }
  });

export default orderReducer;
