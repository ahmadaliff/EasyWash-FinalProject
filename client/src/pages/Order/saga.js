import { call, put, takeLatest } from 'redux-saga/effects';
import toast from 'react-hot-toast';

import intlHelper from '@utils/intlHelper';

import { apiCreateOrder } from '@domain/api';

import { setLoading, showPopup } from '@containers/App/actions';
import { CREATE_ORDER } from '@pages/Order/constants';

function* sagaCreateOrder({ data, callback }) {
  yield put(setLoading(true));
  try {
    const response = yield call(apiCreateOrder, data);
    toast.success(intlHelper({ message: response.message }));
    yield call(callback);
  } catch (error) {
    if (error?.response?.status === 400 || error?.response?.status === 404) {
      toast.error(intlHelper({ message: error.response.data.message }));
    } else {
      yield put(showPopup());
    }
  }
  yield put(setLoading(false));
}

export default function* orderSaga() {
  yield takeLatest(CREATE_ORDER, sagaCreateOrder);
}
