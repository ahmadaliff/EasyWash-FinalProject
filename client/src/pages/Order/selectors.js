import { createSelector } from 'reselect';
import { initialState } from '@pages/Order/reducer';

const selectOrderState = (state) => state.order || initialState;

export const selectOrder = createSelector(selectOrderState, (state) => state.order);
