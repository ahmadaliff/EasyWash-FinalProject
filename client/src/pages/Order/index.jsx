import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { FormattedMessage, FormattedNumber } from 'react-intl';

import MapLeaflet from '@components/MapLeaflet';
import NoData from '@components/NoData';

import { ArrowBack, DryCleaning } from '@mui/icons-material';
import { Button } from '@mui/material';

import { selectOrder } from '@pages/Order/selectors';
import { actionCreateOrder, actionResetOrder } from '@pages/Order/action';

import classes from '@pages/Order/style.module.scss';
import { selectUser } from '@containers/Client/selectors';

const Order = ({ user, orders }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [location, setLocation] = useState();

  useEffect(() => {
    user?.role !== 'user' && navigate(-1);
    return () => {
      if (orders) dispatch(actionResetOrder());
    };
  }, [dispatch, navigate, orders, user]);

  const handleLocation = (loc) => {
    setLocation(loc);
  };

  const getTotalPrice = () => {
    let totalPrice = 0;
    orders?.forEach((order) => {
      totalPrice += order.servicePrice * order.quantity;
    });
    return totalPrice;
  };

  const handleOrder = () => {
    const orderItems = orders.map(({ serviceName, servicePrice, createdAt, updatedAt, userId, ...order }) => order);
    dispatch(actionCreateOrder({ location: JSON.stringify(location), orderItems }, () => navigate('/user/order')));
  };
  return (
    <main className={classes.mainWrap}>
      <div className={classes.header}>
        <h3>
          <FormattedMessage id="app_detail_order" />
        </h3>
        <Button
          variant="contained"
          className={classes.backButton}
          startIcon={<ArrowBack />}
          onClick={() => navigate(-1)}
        >
          <FormattedMessage id="app_back" />
        </Button>
      </div>
      <div className={classes.inputLabel}>
        <FormattedMessage id="app_location_order" />
        <MapLeaflet handleLocation={handleLocation} permanent={false} />
      </div>

      <div className={classes.inputLabel}>
        <FormattedMessage id="app_detail_order" />
      </div>
      {orders?.map((order, key) => (
        <div className={classes.service} key={key}>
          <DryCleaning />
          <div className={classes.serviceInfoWrap}>
            <hr />
            <div className={classes.serviceInfo}>
              <p>
                {order?.serviceName} x {order?.quantity}
              </p>
              <p>
                <FormattedNumber
                  // eslint-disable-next-line no-unsafe-optional-chaining
                  value={order?.servicePrice * order?.quantity}
                  // eslint-disable-next-line react/style-prop-object
                  style="currency"
                  currency="IDR"
                  minimumFractionDigits={0}
                />
              </p>
            </div>
            <hr />
          </div>
        </div>
      ))}
      {(orders?.length === 0 || !orders) && <NoData />}
      <div className={classes.totalPrice}>
        <p>
          <FormattedMessage id="app_total_price" />
        </p>
        <p>
          <FormattedNumber
            value={getTotalPrice()}
            // eslint-disable-next-line react/style-prop-object
            style="currency"
            currency="IDR"
            minimumFractionDigits={0}
          />
        </p>
      </div>

      <Button
        variant="contained"
        className={classes.createButton}
        onClick={handleOrder}
        disabled={orders?.length === 0 || !orders}
      >
        <FormattedMessage id="app_order" />
      </Button>
    </main>
  );
};

Order.propTypes = {
  orders: PropTypes.array,
  user: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  orders: selectOrder,
  user: selectUser,
});
export default connect(mapStateToProps)(Order);
