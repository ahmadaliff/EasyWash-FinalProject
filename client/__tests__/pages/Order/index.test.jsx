import Order from '@pages/Order';
import { render, fireEvent, act } from '@utils/testHelper';

let wrapper;
const mockNavigate = jest.fn();
const mockDispatch = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => mockDispatch,
}));
jest.mock('@pages/Order/selectors', () => ({
  selectOrder: () => [{ serviceName: 'reguler', quantity: 2, servicePrice: 2000 }],
}));
jest.mock('@components/MapLeaflet', () => jest.fn().mockImplementation(() => <div />));

beforeEach(() => {
  act(() => {
    wrapper = render(<Order />);
  });
});

describe('Order Page', () => {
  test('render', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('order-page')).toBeInTheDocument();
  });
  test('navigate back when back button click', async () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-back');
    await act(async () => {
      fireEvent.click(button);
    });
    expect(mockNavigate).toHaveBeenCalledWith(-1);
  });
  test('should be call func handle order', async () => {
    const { getByTestId } = wrapper;
    const button = getByTestId('button-order');
    await act(async () => {
      fireEvent.click(button);
    });
    expect(mockDispatch).toHaveBeenCalled();
  });
  test('Matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
