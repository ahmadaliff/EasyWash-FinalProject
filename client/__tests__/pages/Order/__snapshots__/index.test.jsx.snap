// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Page Matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="mainWrap"
        data-testid="order-page"
      >
        <div
          class="header"
        >
          <h3>
            Detail pesanan
          </h3>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium backButton css-sghohy-MuiButtonBase-root-MuiButton-root"
            data-testid="button-back"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ArrowBackIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                />
              </svg>
            </span>
            Kembali
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="inputLabel"
        >
          Lokasi Penjemputan
          <div />
        </div>
        <div
          class="inputLabel"
        >
          Detail pesanan
        </div>
        <div
          class="service"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DryCleaningIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.56 11.36 13 8.44V7c0-.55-.45-1-1-1s-1-.45-1-1 .45-1 1-1 1 .45 1 1h2c0-1.84-1.66-3.3-3.56-2.95-1.18.22-2.15 1.17-2.38 2.35-.3 1.56.6 2.94 1.94 3.42v.63l-6.56 2.92c-.88.38-1.44 1.25-1.44 2.2v.01C3 14.92 4.08 16 5.42 16H7v6h10v-6h1.58c1.34 0 2.42-1.08 2.42-2.42v-.01c0-.95-.56-1.82-1.44-2.21zM18.58 14H17v-1H7v1H5.42c-.23 0-.42-.19-.42-.43 0-.17.1-.32.25-.38l6.75-3 6.75 3c.15.07.25.22.25.39 0 .23-.19.42-.42.42z"
            />
          </svg>
          <div
            class="serviceInfoWrap"
          >
            <hr />
            <div
              class="serviceInfo"
            >
              <p>
                reguler
                 x 
                2
              </p>
              <p>
                Rp 4.000
              </p>
            </div>
            <hr />
          </div>
        </div>
        <div
          class="totalPrice"
        >
          <p>
            Harga Total
          </p>
          <p>
            Rp 4.000
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium createButton css-sghohy-MuiButtonBase-root-MuiButton-root"
          data-testid="button-order"
          tabindex="0"
          type="button"
        >
          Pesan
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="mainWrap"
      data-testid="order-page"
    >
      <div
        class="header"
      >
        <h3>
          Detail pesanan
        </h3>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium backButton css-sghohy-MuiButtonBase-root-MuiButton-root"
          data-testid="button-back"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ArrowBackIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
              />
            </svg>
          </span>
          Kembali
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="inputLabel"
      >
        Lokasi Penjemputan
        <div />
      </div>
      <div
        class="inputLabel"
      >
        Detail pesanan
      </div>
      <div
        class="service"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="DryCleaningIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19.56 11.36 13 8.44V7c0-.55-.45-1-1-1s-1-.45-1-1 .45-1 1-1 1 .45 1 1h2c0-1.84-1.66-3.3-3.56-2.95-1.18.22-2.15 1.17-2.38 2.35-.3 1.56.6 2.94 1.94 3.42v.63l-6.56 2.92c-.88.38-1.44 1.25-1.44 2.2v.01C3 14.92 4.08 16 5.42 16H7v6h10v-6h1.58c1.34 0 2.42-1.08 2.42-2.42v-.01c0-.95-.56-1.82-1.44-2.21zM18.58 14H17v-1H7v1H5.42c-.23 0-.42-.19-.42-.43 0-.17.1-.32.25-.38l6.75-3 6.75 3c.15.07.25.22.25.39 0 .23-.19.42-.42.42z"
          />
        </svg>
        <div
          class="serviceInfoWrap"
        >
          <hr />
          <div
            class="serviceInfo"
          >
            <p>
              reguler
               x 
              2
            </p>
            <p>
              Rp 4.000
            </p>
          </div>
          <hr />
        </div>
      </div>
      <div
        class="totalPrice"
      >
        <p>
          Harga Total
        </p>
        <p>
          Rp 4.000
        </p>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium createButton css-sghohy-MuiButtonBase-root-MuiButton-root"
        data-testid="button-order"
        tabindex="0"
        type="button"
      >
        Pesan
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
